db/push.go

```diff
@@ -0,0 +1,96 @@
+package db
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	nomsjson "roci.dev/diff-server/util/noms/json"
+)
+
+// These request/response types might want to live in a different location, eg 
+// in diff-server/serve/types.
+type Mutation struct {
+	ID   uint64          `json:"id"`
+	Name string          `json:"name"`
+	Args json.RawMessage `json:"args"`
+}
+
+type BatchPushRequest struct {
+	ClientID  string     `json:"clientId"`
+	Mutations []Mutation `json:"mutations"`
+}
+
+type MutationInfo struct {
+	ID    string `json:"id"`
+	Error string `json:"error"`
+}
+
+type BatchPushResponse struct {
+	MutationInfos []MutationInfo `json:"mutationInfos"`
+}
+
+type BatchPushInfo struct {
+	HTTPStatusCode    int               `json:"httpStatusCode"`
+	ErrorMessage      string            `json:"errorMessage"`
+	BatchPushResponse BatchPushResponse `json:"batchPushResponse"`
+}
+
+func push(pending []Local, url string, dataLayerAuth string, obfuscatedClientID string) (BatchPushInfo, error) {
+	var req BatchPushRequest
+	req.ClientID = obfuscatedClientID
+	for _, p := range pending {
+		var args bytes.Buffer
+		if err := nomsjson.ToJSON(p.Args, &args); err != nil {
+			return BatchPushInfo{}, err
+		}
+		req.Mutations = append(req.Mutations, Mutation{p.MutationID, p.Name, args.Bytes()})
+	}
+	reqBody, err := json.Marshal(req)
+	if err != nil {
+		return BatchPushInfo{}, err
+	}
+
+	httpReq, err := http.NewRequest("POST", url, bytes.NewReader(reqBody))
+	if err != nil {
+		return BatchPushInfo{}, err
+	}
+	httpReq.Header.Add("Authorization", dataLayerAuth)
+	httpResp, err := http.DefaultClient.Do(httpReq)
+	if err != nil {
+		return BatchPushInfo{}, err
+	}
+
+	var info BatchPushInfo
+	info.HTTPStatusCode = httpResp.StatusCode
+	if httpResp.StatusCode == http.StatusOK {
+		var resp BatchPushResponse
+		if err := json.NewDecoder(httpResp.Body).Decode(&resp); err != nil {
+			return info, fmt.Errorf("error decoding batch push response: %w", err)
+		}
+		info.BatchPushResponse = resp
+	} else {
+		body, err := ioutil.ReadAll(httpResp.Body)
+		var s string
+		if err == nil {
+			s = string(body)
+		} else {
+			s = err.Error()
+		}
+		info.ErrorMessage = s
+	}
+
+	return info, nil
+}
+
+type pusher interface {
+	Push(pending []Local, url string, dataLayerAuth string, obfuscatedClientID string) (BatchPushInfo, error)
+}
+
+type defaultPusher struct{}
+
+func (defaultPusher) Push(pending []Local, url string, dataLayerAuth string, obfuscatedClientID string) (BatchPushInfo, error) {
```

Comment:
Curious why not just inline `push` here? Is there a need for it to exist separately as a bare function?